Clean-up use of guards

romainbrenguier opened the pull request:
In some places, guards are passed around but never really used, in other places they can be replaced by expressions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on src/pointer-analysis/goto_program_dereference.h:
```diff
@@ -19,6 +19,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "value_sets.h"
 #include "value_set_dereference.h"
 
+class guardt;
```
I'm not exactly sure about the order of the commits here, but at the very end this shouldn't be needed anymore at all I think?

romainbrenguier replied:
It is still used in some unused deprecated function and I'm not sure of what to do with all these functions (most of the code in this class is unused).

tautschnig replied:
Remove the cruft. `git revert` is not all that hard.

romainbrenguier replied:
I added a commit to replace that argument by an exprt. I also updated the date of commits so that they appear in the right order.

romainbrenguier committed:
Make dereference non-virtual

dereference is not overridden, we add final to value_set_dereferencet to
be sure of that.
Calls to virtual functions are slow, and in general it is better to
avoid virtual recursive functions (like derefence was).

romainbrenguier committed:
Remove unused guard argument from dereference

This argument was never actually used.

romainbrenguier committed:
Remove unused guard argument of dereference_rec

This is never actually used. This allows to remove a lot of useless
operations.

romainbrenguier committed:
Remove unused include of guard.h

The include statement is not usefull anymore, as the usage of guard
disappeared from this file.

romainbrenguier committed:
Replace guardt by exprt::operandst in rw_set

This was not making really use of guardt, since no guardt specific
operations where performed, but instead just represents a conjunction.
We replace that by a list of expressions and do the conjunction when
needed.
This avoids dependencies on guardt.

romainbrenguier committed:
Remove unused code from goto_program_dereference

A lot of code in goto_program_dereference was deprecated, commented or
documented as unused. This commit remove these parts of the code.

danpoe commented on src/goto-instrument/rw_set.cpp:
```diff
@@ -79,14 +79,15 @@ void _rw_set_loct::compute()
+  bool w,
   const std::string &suffix,
-  const guardt &guard)
+  const exprt::operandst &guard)
```
Can we rename the `guard` parameters maybe to `guard_conjuncts`? Otherwise it might be surprising to find a parameter named `guard` of type `exprt::operandst`.

danpoe reviewed (approved):
LGTM

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 18f4ba54942da101d0eeccd97657b52c81dcd744).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100961989
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: af6bd9c9a39876bf6ce61ca4d5641d224c2ece7a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100962569
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c459861edb75ded396371425a950d8727482119c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100964248

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cee38c52bce857d6b172d526e6cb17f740bc8b9c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101067140
